kyo-prelude/shared/src/main/scala/kyo/Kyo.scala

```diff
@@ -15,121 +15,202 @@ object Kyo:
         v1.map(t1 => v2.map(t2 => v3.map(t3 => v4.map(t4 => (t1, t2, t3, t4)))))
 
-    object seq:
-        def map[A, B, S, S2](seq: Seq[A])(f: A => B < S2)(using Frame): Seq[B] < (S & S2) =
-            seq.size match
-                case 0 => Seq.empty
-                case 1 => f(seq(0)).map(Seq(_))
-                case size =>
-                    seq match
-                        case seq: IndexedSeq[A] =>
-                            Loop.indexed(Chunk.empty[B]) { (idx, acc) =>
-                                if idx == size then Loop.done(acc.toSeq)
-                                else f(seq(idx)).map(u => Loop.continue(acc.append(u)))
-                            }
-                        case seq: List[A] =>
-                            Loop(seq, Chunk.empty[B]) { (seq, acc) =>
-                                seq match
-                                    case Nil          => Loop.done(acc.toSeq)
-                                    case head :: tail => f(head).map(u => Loop.continue(tail, acc.append(u)))
-                            }
-                        case seq =>
-                            Chunk.from(seq).map(f).map(_.toSeq)
-                    end match
-
-        def foreach[A, B, S](seq: Seq[A])(f: A => Unit < S)(using Frame): Unit < S =
-            seq.size match
-                case 0 =>
-                case 1 => f(seq(0))
-                case size =>
-                    seq match
-                        case seq: IndexedSeq[A] =>
-                            Loop.indexed { idx =>
-                                if idx == size then Loop.done
-                                else f(seq(idx)).andThen(Loop.continue)
-                            }
-                        case seq: List[A] =>
-                            Loop(seq) {
+    def foreach[A, B, S, S2](seq: Seq[A])(f: A => B < S2)(using Frame): Chunk[B] < (S & S2) =
+        seq.size match
+            case 0 => Chunk.empty
+            case 1 => f(seq(0)).map(Chunk(_))
+            case size =>
+                seq match
+                    case seq: List[A] =>
+                        Loop(seq, Chunk.empty[B]) { (seq, acc) =>
+                            seq match
+                                case Nil          => Loop.done(acc)
+                                case head :: tail => f(head).map(u => Loop.continue(tail, acc.append(u)))
+                        }
+                    case seq =>
+                        val indexed = toIndexed(seq)
+                        Loop.indexed(Chunk.empty[B]) { (idx, acc) =>
+                            if idx == size then Loop.done(acc)
+                            else f(indexed(idx)).map(u => Loop.continue(acc.append(u)))
+                        }
+                end match
+
+    def foreachDiscard[A, B, S](seq: Seq[A])(f: A => Unit < S)(using Frame): Unit < S =
+        seq.size match
```

Comment:
I think you want to use `knownSize` to avoid iterating `List`s.

Reply:
thanks! I didn't know `knownSize`